README.md
# DataDog log export

Download logs from DataDog via the API and save them to your PC.\
This tool originated from an existent GitHub Project: [datadog-downloader](https://github.com/wegift/datadog-downloader) by [WeGift](https://github.com/wegift), a script that allows you to download a large number of logs matching a particular query rather than being bound by the limit of 5000 imposed on the export button in the UI.

From this base, a complete tool has been developed to automate the process, giving the user the largest flexibility in terms of input and output.
Logs are output as a file, and the format is one of the parameters (the choice is between `JSON` and `CSV`).

Fields that will be used to create the result file are fully customizable, they are one of the parameters, mandatory of course, see forward for details.

## Usage

The downloader is a container that exposes two endpoints:

- `POST http://[host]:[port]/export` for sending a download request, parameters will be passed via the `JSON` payload, if nothing goes wrong it returns the URL to call in order to download the file.

The input JSON must follow these rules:

|  Parameter  |  Mandatory  |  Description  | Allowed values  | Default Value  |
|---|---|---|---|---|
| query | YES  | The DataDog filter query. Take care when quoting on the command line, single quote the entire query for best results.  | `datadog filter string` | `none` |
| columns | YES | A list of DataDog logs attributes to include in the export, with label to assign in the export file and a default value | `JSON object` | `none`  |
| from | NO | Start date/time | `string date` | `1 day ago` |
| to | NO | End date/time | `string date` | `current timestamp` |
| pageSize | NO | How many results to download at a time, default 1000 limit of 5000 | `[1..5000]` | `1000` |
| outputFormat | NO | Format of file to write results to | `['JSON','CSV']` | `JSON` |
| outputFile | NO | Path of file to write results to (the extension is related to outputFormat) | `file name (without extension)` | `results` |
| verbose |  NO | Flag for having tool's activities logged on console | `true\|false` | `false` |

Note: Date/times are parsed by JS `Date` constructor, e.g. 2022-01-01.

The result is an `HTTP 202` status code, the body contains a link for downloading the file, this link has a validity of 36 hours, then the file will be deleted.

- `GET http://[host]:[port]/{result_file}` for getting back the exported log file. 

If the export is still in progress, you will get yet a `HTTP 202` status, if an error occurred during the export phase an `HTTP 400` will be returned, otherwise if `result_file` exists and it was created within the previous 36 hours, it will be downloaded, otherwise, you will get an `HTTP 404` error.

## How does it work

The user requests the export, the tool launches a separate process to export and create a file, and responds immediately to the user with the link, the HTTP STATUS sent is `202`. In the bucket is created immediately a `sentinel file` to represent export was started and is in progress, if any error occurred the sentinel file will be deleted and an `error file` will be created.

Both `sentinel` and `error` files have the same name as the `export` one, with the suffixes `.sem` and `.ERR` respectively.

The endpoint `GET /{filename}` will check the presence of these files, in precise order:

1. `{filename}.sem`: if the sentinel is present, the export is still in progress, a `HTTP 202` is returned, wait and retry later.
2. `{filename}.ERR`: an error occurred, a `HTTP 400` is returned.
3. `{filename}`: the sentinel was deleted, the error file is not present, so the export file may be present. If all goes right, a `HTTP 200` will be returned, and the file will be send to the client.

## Storing result files

Once created, a report file will be stored for the next 36 hours in a Cloud Storage Bucket, according to container's environment configuration.

The allowed Cloud storage providers actually are `Google` and `AWS`

## Output file's columns

Which logs data put in the export file?

The list of attributes of a Datadog log is very large and it depends on which logs channel we wants to analyze and **what** we wants to analyze, so the best answer to the question is `it depends!`.\
You have the power, you can choice what you want to have in the file.\
The log's timestamp is the only attribute that is in all results, as first information.

You are free to add whatever you want, using the `columns` attribute of the JSON payload of the command input; it's possible to define the `label` (how the information will be named in the file) and its `path` inside the Datadog logs, and optionally you can set a default value in case the information is not present in a single log entry (an empty string `""` as system default value).

The _root_ of log attributes is `attributes`, so (for example) the timestamp is referred as `attributes.timestamp`

The `columns` input attribute is an array of tuples containing label, attribute's path and optional default value if the attribute isn't found.\
The `path` must be written as a string, with an hashtag (`#`) as separator (come back to timestamp example, it should be added as `attributes#timestamp`)

## Authentication

You will need an API key and an app key to access the DataDog api.
These should be provided in environment variables as seen above.

API keys are global for a DataDog account and can be found in organization settings.
App keys are personal to your profile and can be generated in personal settings.

## Environment variables

| Variable name  | Mandatory  | Description  | Allowed values  | Example value  |
|---|---|---|---|---|
| PORT | YES | TCP Port Server puts itself to listening on | `1 .. 60000` | `8080` |
| DD_SITE | YES | DataDog site from which logs will be downloaded | `URL \| hostname` | `datadoghq.com` |
| DD_API_KEY | YES | DataDog API KEY | `API KEY string` | `aaabbbcccdddeeefff` |
| DD_APP_KEY | YES | DataDog Application KEY | `APP KEY string` | `xxxxyyyyzzzzwwww` |
| STORAGE_TYPE | YES | Cloud Storage Provider to store result files | `['GOOGLE'\,'AWS']` | `AWS`
| GOOGLE_APPLICATION_CREDENTIALS | `if STORAGE_TYPE is GOOGLE` | Path of Google's JSON credentials file | `path to json file` | `/path/to/google_credentials.json` |
| GOOGLE_BUCKET_NAME | `if STORAGE_TYPE is GOOGLE` | Bucket name to store data | `string` | `datadog_storage` |
| AWS_ACCESS_KEY_ID | `if STORAGE_TYPE is AWS` | Access Key for AWS account | `Access Key string ` | `xxxxaaaazzzzbbbbyyy` |
| AWS_SECRET_ACCESS_KEY | `if STORAGE_TYPE is AWS` | Secret for AWS account | `Secret string ` | `xxxx-aaaaz$zzzbbbbyyy` |
| AWS_BUCKET_NAME | `if STORAGE_TYPE is AWS` | Bucket name to store data | `string` | `datadog_storage` |
| AWS_REGION | `if STORAGE_TYPE is AWS` | AWS region where bucket is located | `AWS region` | `eu-north-1` |

## Local Dev

Run `npm install`.

Copy `.env.example` to `.env.production` and add a valid data.

## Run

`node index.js` to start the server (also a `npm start` does the job).

Do you want to asking for all HTTP 500 errors on `/api/*` calls on 27th March between 12 and 13 GMT, saving a CSV file, with `date, context.request.uri, context.request.headers.x-myown-header` as output, with verbose logging to console? Type

```
curl -H "Content-type: application/json" -X POST "http://localhost:8080/export" \
-d'{"query":"@context.request.uri:\/api\/* @http.status_code:500","verbose":true,"from":"2023-03-27T12:00:00.000Z","to":"2023-03-27T13:00:00.000Z","outputFormat":"csv","columns":[{"label":"URI","path":"attributes#context#request#uri"},{"label":"My Header","path":"attributes#context#request#headers#x-myown-header"}]}'
```

You will get `http://localhost:8080/1680087684088_results.csv` as response, or something similar, the prefix is timestamp of request, used for cleaning archives after the limit of 36 hours.

Calling the URL from a browser will force to download the file, otherwise you can use cUrl (according previous request)

`curl http://localhost:8080/1680087684088_results.csv -o "1680087684088_results.csv"`

to have the file stored in your computer.

## Test

The tool uses `vitest` as test suite. Run `vitest` to launch the tests, or `npm run start-dev`.


CHANGELOG.md
# Changelog

## Rel. 0.4.0 [2023-04-14]

Making exported file's columns customizable, as a new input arguments

## Rel. 0.3.0 [2023-04-11]

Adding test functionality

## Rel. 0.2.0 [2023-04-04]

Making request and export process separate and asynchronous. Use of sentinel and error files.

## Rel. 0.1.0 [2023-03-31]

First release, the tool works! Listening on defined port and exposes two endpoints:

- `POST /export` for requesting a datadog export, the output is the link that can be used to retrieve the file
- `GET /{filename}` for saving the export generated file

tests/index.test.js
import * as dotenv from "dotenv";
dotenv.config({ path: `.env.${process.env.NODE_ENV}` });
import { describe, expect, test } from 'vitest';
import * as Index from '../index';
import * as storage from '../storage/testStorage';
import * as fs from "fs";

describe("Cloud storage interface test", () => {

    test("Successful upload file to storage bucket", () => {
        let result;
        storage.uploadFile('/path/to/file', 'destPathFile', () => { result = false }, () => { result = true });
        expect(result).toBeTruthy();
    })

    test("Failing upload file to storage bucket", () => {
        let result;
        storage.uploadFile('/path/to/failing/file', 'destPathFile', () => { result = false }, () => { result = true });
        expect(result).toBeFalsy();
    })

});

describe("Endpoint test", () => {

    test("Trying getting file with export process still in progress", () => {
        fs.writeFileSync('./test', 'test');
        fs.writeFileSync('./test.sem', 'test');
        Index.getFile('test', (data) => {
            expect(data.status).toEqual(202);
            fs.unlinkSync('./test');
            fs.unlinkSync('./test.sem');
        });
    });

    test("Failing getting file with error file", () => {
        fs.writeFileSync('./test1', 'test');
        fs.writeFileSync('./test1.ERR', 'test');
        Index.getFile('test1', (data) => {
            expect(data.status).toEqual(404);
            fs.unlinkSync('./test1');
            fs.unlinkSync('./test1.ERR');
        });
    });

    test("Getting file", () => {
        fs.writeFileSync('./test2', 'test');
        Index.getFile('test2', (data) => {
            expect(data.status).toEqual(200);
            fs.unlinkSync('./test2');
        });
    });

    test("Failing getting non existent file", () => {
        Index.getFile('test_to_fail', (data) => {
            expect(data.status).toEqual(404);
        });
    });

});


server.js
import * as dotenv from "dotenv";
dotenv.config({ path: `.env.${process.env.NODE_ENV || 'production'}` });

import express from "express";
import * as Base from "./index.js";

const PORT = process.env.PORT || 8080;
const HOST = '0.0.0.0';

const app = express();

app.use(express.urlencoded({extended: true}));
app.use(express.json());

app.get('/:filename', (req, res) => {
  Base.getFile(req.params.filename, (response) => {
    if (response.status == 200) {
      res.sendFile(response.filePath);
    } else {
      res.status(response.status).send(response.message);
    }
  });
});

app.post('/export', (req, res) => {
  req.body.baseURL = req.protocol + '://' + req.get('host');
  Base.exportRequest(req.body, (response) => {
    res.status(response.status).send(response.message);
  });
});

app.listen(PORT, HOST, () => {
  console.log(`Running on http://${HOST}:${PORT}`);
});

export_log.js
import { v2 } from "@datadog/datadog-api-client";
import * as fs from "fs";

const configuration = v2.createConfiguration();
const apiInstance = new v2.LogsApi(configuration);

import { storage } from "./storage/storage.js";

const verbose = process.env.verbose ?? false;
const outputFile = process.env.outputFile;
const outputFormat = process.env.outputFormat;
const destFileName = process.env.destFileName;

let initialParams = {};

function yesterday() {
    return new Date(new Date() - 86400000);
}

const startTime = new Date();

const columns = JSON.parse(process.env.columns);

/* 
    Performs a deep walk into an object to see if a specific key (or sub key) exists:
    - if the key was found, it returns the value
    - if the key doesn't exists, it returns null
*/
function deepCheck(rif, key) {
    let k = key.shift();
    if (rif[k] != undefined) {
        if (key.length == 0) {
            return rif[k];
        } else {
            return deepCheck(rif[k], key);
        }
    } else {
        return null;
    }
}

initialParams = {
    filterQuery: process.env.query,
    filterFrom: process.env.from ? new Date(process.env.from) : yesterday(),
    filterTo: process.env.to ? new Date(process.env.to) : new Date(),
    pageLimit: process.env.pageSize ? Math.min(process.env.pageSize, 5000) : 1000,
};

if (verbose) {
    console.log(destFileName + " *** Downloading logs:\n" + JSON.stringify(initialParams, null, 2) + "\noutputFormat: " + outputFormat + "\noutputFile: " + outputFile + "\nVerbose: " + verbose.toString() + "\n");
    console.log(destFileName + ' *** Start at ' + startTime);
}

getLogs(apiInstance, initialParams, (data) => {

    if (storage != null) {
        storage.uploadFile(outputFile, destFileName,
            (e) => {
                if (verbose) {
                    console.log(destFileName + ' *** Error occurred: ' + e);
                }
                storage.uploadFile('./semaphore', destFileName + '.ERR',
                    (e) => { },
                    () => {
                        storage.deleteFile(destFileName + '.sem');
                    }
                );
            },
            () => {
                fs.unlinkSync(outputFile, () => { });
                storage.deleteFile(destFileName + '.sem');
            }
        );
    } else {
        storage.uploadFile('./semaphore', destFileName + '.ERR',
            (e) => { },
            () => {
                storage.deleteFile(destFileName + '.sem');
            }
        );
    }

});



async function getLogs(apiInstance, params, cb) {
    let data = [];
    let contentLen = 0;
    let nextPage = null;
    let n = 0;
    try {
        if (outputFormat == "CSV") {
            let head = 'date';
            columns.forEach(c => { head += ',' + c.label });
            fs.writeFileSync(outputFile, head + "\n");
        }
        do {
            if (verbose) {
                console.log(`${destFileName} *** Requesting page ${++n}`);
            }
            const query = nextPage ? { ...params, pageCursor: nextPage } : params;
            const result = await apiInstance.listLogsGet(query);
            let csvData = "";
            result.data.forEach(tmpRow => {
                if (outputFormat == "CSV") {
                    csvData += '"' + (new Date(tmpRow.attributes.timestamp).toISOString()) + '"';
                }
                if (outputFormat == "JSON") {
                    let tmpData = { date: tmpRow.attributes.timestamp }
                }
                columns.forEach(c => {
                    let attrs = c.path.split('#');
                    let value = deepCheck(tmpRow.attributes, attrs);
                    if (typeof value == 'object') {
                        value = JSON.stringify(value);
                    }
                    if (value == null) {
                        value = c.default ? c.default.toString() : '';
                    }
                    if (outputFormat == "CSV") {
                        csvData += ',"' + value + '"';
                    }
                    if (outputFormat == "JSON") {
                        tmpData[c.label] = value;
                    }
                });
                if (outputFormat == "CSV") {
                    csvData += '\n';
                }
                if (outputFormat == "JSON") {
                    data.push(tmpData);
                }
            });
            if (outputFormat == "CSV") {
                fs.appendFileSync(outputFile, csvData);
            }
            contentLen += result.data.length;
            if (verbose) {
                console.log(`${destFileName} *** Writing ${result.data.length} logs to ${outputFile} (${contentLen} total)`);
            }
            nextPage = result?.meta?.page?.after;
        } while (nextPage);
        if (outputFormat == "JSON") {
            fs.writeFileSync(outputFile, JSON.stringify(data, null, 2));
        }
    } catch (e) {
        if (verbose) {
            console.log(destFileName + " *** An error occurred: " + e);
            fs.writeFileSync('./error', e.toString());
            storage.uploadFile('./error', destFileName + '.ERR',
                (e) => {
                    fs.unlinkSync('./error');
                },
                () => {
                    fs.unlinkSync('./error');
                    storage.deleteFile(destFileName + '.sem');
                }
            );
        }
    }
    if (verbose) {
        console.log(destFileName + " *** Process ended.");
    }
    cb(contentLen);
}

storage/aws.js
import * as dotenv from "dotenv";
import * as fs from "fs";
import { GetObjectCommand, PutObjectCommand, DeleteObjectCommand, S3Client, ListObjectsV2Command } from "@aws-sdk/client-s3";

dotenv.config({ path: `.env.${process.env.NODE_ENV || 'production'}` });

const client = new S3Client({})

async function uploadFile(filePath, destFileName, fail, success) {
    let data = fs.readFileSync(filePath);

    const command = new PutObjectCommand({
        Bucket: process.env.AWS_BUCKET_NAME,
        Key: destFileName,
        Body: Buffer.from(data)
    });

    try {
        const response = await client.send(command);
        success();
    } catch (err) {
        fail(err);
    }
};

async function purgeOlderFiles(fileName) {
    const command = new ListObjectsV2Command({
        Bucket: process.env.AWS_BUCKET_NAME,
        MaxKeys: 20,
    });
    let fileExists = false;
    try {
        let isTruncated = true;
        // List all files in bucket
        while (isTruncated) {
            const { Contents, IsTruncated, NextContinuationToken } = await client.send(command);
            Contents.forEach(file => {
                if (file.Key.split('_').length > 1) {
                    // Delete files older than 36 hours
                    let fileTime = file.Key.split('_')[0];
                    let expireTime = fileTime * 1 + 60 * 60 * 36 * 1000;
                    let now = new Date() * 1;
                    if (expireTime < now) {
                        deleteFile(file.Key);
                    }
                    // If file exists and it isn't expired we can proceed
                    if (file.Key == fileName && expireTime > now) {
                        fileExists = true;
                    }
                }
            })
            isTruncated = IsTruncated;
            command.input.ContinuationToken = NextContinuationToken;
        }
    } catch (err) {
        fileExists = false;
    }
    return fileExists;
}

async function deleteFile(fileName) {
    const command = new DeleteObjectCommand({
        Bucket: process.env.AWS_BUCKET_NAME,
        Key: fileName
    });

    try {
        await client.send(command);
    } catch (err) {
        console.error(err);
    }
}

async function downloadFile(fileName, fail, success) {
    const command = new GetObjectCommand({
        Bucket: process.env.AWS_BUCKET_NAME,
        Key: fileName
    });

    try {
        // Downloads the file if it exists
        if (true === await purgeOlderFiles(fileName)) {
            const response = await client.send(command);
            const str = await response.Body.transformToByteArray();
            fs.writeFileSync('/tmp/' + fileName, Buffer.from(str));
            success('/tmp/' + fileName);
        } else {
            fail('File not found');
        }
    } catch (err) {
        fail(err);
    }
};

export { uploadFile, downloadFile, deleteFile }

storage/gcp.js
import * as dotenv from "dotenv";

dotenv.config({ path: `.env.${process.env.NODE_ENV || 'production'}` });

// Imports the Google Cloud client library
import { Storage } from "@google-cloud/storage";

// Creates a client
const storage = new Storage();

const bucketName = process.env.GOOGLE_BUCKET_NAME;

async function uploadFile(filePath, destFileName, fail, success) {
    const options = {
        destination: destFileName,
    };

    await storage.bucket(bucketName).upload(filePath, options);
    success();
}

async function deleteFile(fileName) {
    await storage.bucket(bucketName).file(fileName).delete();
}

async function purgeOlderFiles(fileName) {
    let fileExists = false;
    // List all files in bucket
    const [files] = await storage.bucket(bucketName).getFiles();
    files.forEach(file => {
        if (file.name.split('_').length > 1) {
            // Delete files older than 36 hours
            let fileTime = file.name.split('_')[0];
            let expireTime = fileTime * 1 + 60 * 60 * 36 * 1000;
            let now = new Date() * 1;
            if (expireTime < now) {
                deleteFile(file.name);
            }
            // If file exists and it isn't expired we can proceed
            if (file.name == fileName && expireTime > now) {
                fileExists = true;
            }
        }
    });
    return fileExists;
}

async function downloadFile(fileName, fail, success) {
    const options = {
        destination: '/tmp/' + fileName,
    };

    // Downloads the file if it exists
    if (true === await purgeOlderFiles(fileName)) {
        await storage.bucket(bucketName).file(fileName).download(options);
        success('/tmp/' + fileName);
    } else {
        fail('File not found');
    }
}

export { uploadFile, downloadFile, deleteFile };

index.js
import * as dotenv from "dotenv";
dotenv.config({ path: `.env.${process.env.NODE_ENV || 'production'}` });

import * as fs from "fs";

import { spawn } from 'child_process';

const allowedOutputTypes = ['JSON', 'CSV', 'SEM', 'ERR'];

import { storage } from "./storage/storage.js";

function getFile(filename, cb) {
    /* Check for older files ready to be deleted */
    fs.readdir('/tmp', (err, files) => {
        if (!err) {
            files.forEach(f => {
                if (allowedOutputTypes.indexOf(f.split('.')[f.split('.').length - 1].toUpperCase()) > -1) {
                    if (f.split('_').length > 1) {
                        let fileTime = f.split('_')[0];
                        let expireTime = fileTime * 1 + 60 * 60 * 36 * 1000;
                        let now = new Date() * 1;
                        if (expireTime < now) {
                            fs.unlink('/tmp/' + f, () => {
                                console.log('Deleting ' + f + ' (expired by ' + (now - expireTime) / 1000 + ' sec.)');
                            });
                        }
                    }
                }
            });
            if (filename.indexOf('/') > -1) {
                cb({ status: 404, message: "" });
            } else {
                if (storage != null) {
                    // First of all, check if semaphore file is yet present
                    // it would mean export in progress
                    storage.downloadFile(filename + '.sem',
                        (e) => {
                            if (e == 'File not found') {
                                // ok, semaphore file is not present, check if there is
                                // the error file too
                                storage.downloadFile(filename + '.ERR',
                                    (e) => {
                                        if (e == 'File not found') {
                                            // ok, we can try to download file
                                            storage.downloadFile(filename,
                                                (e) => {
                                                    cb({ status: 404, message: e });
                                                },
                                                (fn) => {
                                                    cb({ status: 200, filePath: fn });
                                                });
                                        } else {
                                            // other error, return 400
                                            cb({ status: 404, message: e });
                                        }
                                    },
                                    (fn) => {
                                        // There is the error file, returning 400
                                        cb({ status: 404, message: "" });
                                    }
                                );
                            } else {
                                // other error, return 400
                                cb({ status: 404, message: e });
                            }
                        },
                        (fn) => {
                            // The semaphore file is yet present, so export is still in progress
                            // return HTTP 202 Accepted
                            cb({ status: 202, message: 'The export is still in progress, wait and check again later' });
                        }
                    );
                } else {
                    cb({ status: 400, message: 'Wrong Cloud Storage provider' });
                }
            }
        }
    });
}

function exportRequest(reqBody, cb) {
    const outputFormat = (reqBody.outputFormat ?? "json").toUpperCase();
    const outputFile = '/tmp/' + new Date() * 1 + '_' + (reqBody.outputFile ?? "results") + '.' + outputFormat.toLowerCase();
    const destFileName = outputFile.replace('/tmp/', '');
    const downloadURL = reqBody.baseURL + '/' + destFileName;
    const verbose = reqBody.verbose ? true : false;
    const allowedOutputTypes = ["JSON", "CSV"];

    if (!reqBody.columns) {
        cb({ status: 400, message: "Error: No columns supplied, use columns parameter" });
    }

    if (!reqBody.query) {
        cb({ status: 400, message: "Error: No query supplied, use query parameter" });
    }

    if (allowedOutputTypes.indexOf(outputFormat) == -1) {
        cb({ status: 400, message: "Error: Invalid outputFormat value, allowed ones are ['" + allowedOutputTypes.join("', '") + "']" });
    }

    if (storage != null) {
        storage.uploadFile('./semaphore', destFileName + '.sem',
            (e) => {
                cb({ status: 400, message: e.message });
            },
            () => {
                const child = spawn('node', ['./export_log.js'], {
                    stdio: ['ignore', fs.openSync('./out.log', 'a'), fs.openSync('./out.log', 'a')],
                    detached: true,
                    shell: true,
                    env: {
                        query: reqBody.query,
                        from: reqBody.from,
                        to: reqBody.to,
                        pageSize: reqBody.pageSize,
                        outputFile: outputFile,
                        outputFormat: outputFormat,
                        destFileName: destFileName,
                        verbose: verbose,
                        columns: JSON.stringify(reqBody.columns),
                        STORAGE_TYPE: process.env.STORAGE_TYPE,
                        NODE_ENV: process.env.NODE_ENV
                    },
                });
                child.on('error', (err) => { console.log(err) });
                child.unref();
                cb({ status: 202, message: downloadURL });
            }
        );
    } else {
        cb({ status: 400, message: 'Wrong Cloud Storage provider' });
    }

};

export { getFile, exportRequest };